### PR TITLE
adding clear_job_queue clean step that fix pending bios job failure

### PR DESCRIPTION
**Pull request by kpradip001**

### CODE REVIEW

**Comment by kpradip001** [comment]
@cdearborn  I have resolved conflicts, Please review this PR .

File: ironic/drivers/modules/drac/management.py [hunk 1]
```diff
@@ -389,3 +402,67 @@ def known_good_state(self, task):
...
+    @periodics.periodic(
+        spacing=CONF.drac.query_management_config_job_status_interval)
+    def _query_management_config_job_status(self, manager, context):
+        """Periodic task to check the progress of MANAGEMENT config jobs."""
+
+        filters = {'reserved': False, 'maintenance': False}
+        fields = ['driver_internal_info']
+
+        node_list = manager.iter_nodes(fields=fields, filters=filters)
+        for (node_uuid, driver, driver_internal_info) in node_list:
```
Comment (Rachit7194) [inline ironic/drivers/modules/drac/management.py]: If this change approved by chris then, you need to include conductor_group as the third tuple, according to latest master change.

Please refer link : [#L954](https://github.com/openstack/ironic/blob/master/ironic/drivers/modules/drac/raid.py#L954)

Comment (kpradip001) [reply ironic/drivers/modules/drac/management.py]: > If this change approves by chris then, you need to include conductor_group as the third tuple, according to latest master change.
> 
> Please refer link : [#L954](https://github.com/openstack/ironic/blob/master/ironic/drivers/modules/drac/raid.py#L954)

Okay, thanks

File: ironic/conf/drac.py [hunk 2]
```diff
@@ -21,7 +21,14 @@
...
                help=_('Interval (in seconds) between periodic RAID job status '
                       'checks to determine whether the asynchronous RAID '
-                      'configuration was successfully finished or not.'))
+                      'configuration was successfully finished or not.')),
+    cfg.IntOpt('query_management_config_job_status_interval',
+               default=120,
+               min=1,
+               help=_('Interval (in seconds) between periodic management job '
+                      'status checks to determine whether the asynchronous '
+                      'MANAGEMENT configuration was successfully finished '
```
Comment (cdearborn) [inline ironic/conf/drac.py]: Can you make Management lower case in the line above?

File: ironic/drivers/modules/drac/management.py [hunk 3]
```diff
@@ -183,10 +189,17 @@ def set_boot_device(node, device, persistent=False):
     :raises: DracOperationError on an error from python-dracclient.
     """
 
-    drac_job.validate_job_queue(node)
-
     client = drac_common.get_drac_client(node)
 
+    unfinished_jobs = drac_job.list_unfinished_jobs(node)
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/management.py:195]: Can you add a comment describing why this change is necessary?

File: ironic/drivers/modules/drac/management.py [hunk 4]
```diff
@@ -389,3 +402,68 @@ def known_good_state(self, task):
...
+        """Clear the job queue.
+
+        :param task: a TaskManager instance containing the node to act on.
+        :returns: None if it is completed.
+        :raises: DracOperationError on an error from python-dracclient.
+        """
+        node = task.node
+
+        client = drac_common.get_drac_client(node)
+        client.delete_jobs(job_ids=[_CLEAR_JOB_IDS])
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/management.py:420]: Might want to pass no args in the line above for simplicity.

File: ironic/drivers/modules/drac/management.py [hunk 5]
```diff
@@ -183,10 +189,17 @@ def set_boot_device(node, device, persistent=False):
     :raises: DracOperationError on an error from python-dracclient.
     """
 
-    drac_job.validate_job_queue(node)
-
     client = drac_common.get_drac_client(node)
 
+    unfinished_jobs = drac_job.list_unfinished_jobs(node)
+    if unfinished_jobs:
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/management.py:196]: I think we need to change this logic a little:
if there are unfinished Bios jobs and the job queue cleaning step is about to run then delete just the unfinished bios jobs, otherwise, we want to validate the job queue as before.

File: ironic/drivers/modules/drac/management.py [hunk 6]
```diff
@@ -183,10 +189,17 @@ def set_boot_device(node, device, persistent=False):
...
+        bios_jobs = \
+            [job for job in unfinished_jobs if _BIOS_JOB_NAME in job.name]
+        if bios_jobs:
+            client.delete_jobs(job_ids=[_CLEAR_JOB_IDS])
+            node.driver_internal_info['clear_job_substep'] = 'completed'
+            node.save()
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/management.py]: if we do the above, then i don't think we'll need the above 2 lines or the periodic job and associated methods either.

File: ironic/drivers/modules/drac/management.py [hunk 7]
```diff
@@ -389,3 +402,68 @@ def known_good_state(self, task):
...
+        :param task: a TaskManager instance containing the node to act on.
+        :returns: None if it is completed.
+        :raises: DracOperationError on an error from python-dracclient.
+        """
+        node = task.node
+
+        client = drac_common.get_drac_client(node)
+        client.delete_jobs(job_ids=[_CLEAR_JOB_IDS])
+
+    @METRICS.timer('DracRAID._query_management_config_job_status')
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/management.py]: I don't think we need the periodic job or it's associated methods below.

File: ironic/conf/drac.py [hunk 8]
```diff
@@ -21,7 +21,14 @@
                min=1,
                help=_('Interval (in seconds) between periodic RAID job status '
                       'checks to determine whether the asynchronous RAID '
-                      'configuration was successfully finished or not.'))
+                      'configuration was successfully finished or not.')),
+    cfg.IntOpt('query_management_config_job_status_interval',
```
Comment (cdearborn) [inline ironic/conf/drac.py]: If we remove the periodic job, then we won't need this setting.

File: ironic/tests/unit/drivers/modules/drac/test_management.py [hunk 9]
```diff
@@ -445,6 +445,25 @@ def test_set_boot_device_with_invalid_job_queue(
...
+            'id': 'JID_602553293345',
+            'name': 'ConfigBIOS:BIOS.Setup.1-1',
+            'start_time': 'TIME_NOW',
+            'until_time': 'TIME_NA',
+            'message': 'Task successfully scheduled.',
+            'status': 'Scheduled',
+            'percent_complete': 0}
+        bios_job = test_utils.dict_to_namedtuple(values=bios_job_dict)
+        mock_list_unfinished_jobs.return_value = bios_job
+        mock_client.delete_jobs(job_ids=['JID_CLEARALL_FORCE'])
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_management.py]: Instead of the above line, I think we want to call set_boot_device and then assert that delete_jobs was called.

**Review by cdearborn — CHANGES_REQUESTED** [review]
Could you include adding clearing the job queue to the known good cleaning step in this patch?

**Comment by kpradip001** [comment]
Thanks @cdearborn  for the review comments, I will address it and update the PR soon

File: ironic/drivers/modules/drac/management.py (hunk 3 above)
Comment (kpradip001) [reply ironic/drivers/modules/drac/management.py:195]: Done

File: ironic/drivers/modules/drac/management.py (hunk 5 above)
Comment (kpradip001) [reply ironic/drivers/modules/drac/management.py:196]: Done

File: ironic/drivers/modules/drac/management.py (hunk 4 above)
Comment (kpradip001) [reply ironic/drivers/modules/drac/management.py:420]: I think, here we need to keep args as is, because in case of other pending jobs , need this args to clear those pending jobs.

File: ironic/conf/drac.py (hunk 8 above)
Comment (kpradip001) [reply ironic/conf/drac.py]: Done

File: ironic/drivers/modules/drac/management.py (hunk 7 above)
Comment (kpradip001) [reply ironic/drivers/modules/drac/management.py]: I have removed this now.

File: ironic/drivers/modules/drac/management.py (hunk 6 above)
Comment (kpradip001) [reply ironic/drivers/modules/drac/management.py]: I have removed this 2 lines.

File: ironic/conf/drac.py (hunk 2 above)
Comment (kpradip001) [reply ironic/conf/drac.py]: Done

File: ironic/drivers/modules/drac/management.py (hunk 4 above)
Comment (cdearborn) [reply ironic/drivers/modules/drac/management.py:420]: Ah, I see.  This is passing in the FORCE version of the parameter, where the default does not.  Let's keep this as-is.

File: ironic/tests/unit/drivers/modules/drac/test_management.py [hunk 10]
```diff
@@ -423,17 +440,24 @@ def test_set_boot_device_called_with_unknown_boot_mode(
...
+    @mock.patch.object(drac_job, 'list_unfinished_jobs', spec_set=True,
+                       autospec=True)
     def test_set_boot_device_with_invalid_job_queue(
-            self, mock_validate_job_queue, mock_get_drac_client):
+            self, mock_list_unfinished_jobs, mock_validate_job_queue,
+            mock_get_drac_client):
         mock_client = mock.Mock()
         mock_get_drac_client.return_value = mock_client
+
+        mock_list_unfinished_jobs.side_effect = exception.DracOperationError(
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_management.py:494]: I think in this test, we don't want list_unfinished_jobs to raise an exception, so I think the line above and below should be removed.

Comment (cdearborn) [reply ironic/tests/unit/drivers/modules/drac/test_management.py:494]: Let's add another test case that's pretty much the same as this one that tests list_unfinished_jobs raising an exception.

File: ironic/tests/unit/drivers/modules/drac/test_management.py (hunk 9 above)
Comment (cdearborn) [reply ironic/tests/unit/drivers/modules/drac/test_management.py]: Calling delete_jobs on mock_client doesn't test any code because that is a mocked method.  We should call set_boot_device instead, then assert that delete_jobs was called with a parameter of ['JID_CLEARALL_FORCE'].  We should also have pretty much the same asserts that are in the test_set_boot_device test case.

**Review by cdearborn — CHANGES_REQUESTED** [review]
Looks pretty good!  Just had a few comments...

File: ironic/tests/unit/drivers/modules/drac/test_management.py (hunk 9 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/drac/test_management.py]: done

File: ironic/tests/unit/drivers/modules/drac/test_management.py (hunk 10 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/drac/test_management.py:494]: Done

**Comment by kpradip001** [comment]
@cdearborn I have updated PR by addressing review comments of unit tests, please take a look into this.
Thanks.

File: ironic/tests/unit/drivers/modules/drac/test_management.py [hunk 11]
```diff
@@ -423,17 +440,22 @@ def test_set_boot_device_called_with_unknown_boot_mode(
...
         self.assertEqual(0, mock_client.commit_pending_bios_changes.call_count)
 
     @mock.patch.object(drac_job, 'validate_job_queue', spec_set=True,
                        autospec=True)
+    @mock.patch.object(drac_job, 'list_unfinished_jobs', spec_set=True,
+                       autospec=True)
     def test_set_boot_device_with_invalid_job_queue(
-            self, mock_validate_job_queue, mock_get_drac_client):
+            self, mock_list_unfinished_jobs, mock_validate_job_queue,
+            mock_get_drac_client):
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_management.py:460]: I'm not exactly sure how this test case works...  In order for validate_job_queue to be called, list_unfinished_jobs must return a list containing at least one job that is not a BIOS job.  As it's written, I would think that the mocked list_unfinished_jobs would return None, and validate_job_queue would never be called.  Would you mind looking in to this?

**Review by cdearborn — COMMENTED** [review]
Looks pretty good.  I just had one question on a test case.  I can't quite figure out how it works...

File: ironic/tests/unit/drivers/modules/drac/test_management.py (hunk 11 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/drac/test_management.py:460]: Hi Chris, above test case was for invalid job queue which raises error, I have added another positive test case with name test_set_boot_device_with_validate_job_queue to call validate_job_queue based on specified condition.

File: ironic/drivers/modules/drac/management.py [hunk 12]
```diff
@@ -183,10 +186,21 @@ def set_boot_device(node, device, persistent=False):
...
+    # Otherwise, any pending config jobs can cause creating new config jobs
+    # to fail.
+    unfinished_jobs = drac_job.list_unfinished_jobs(node)
+    if unfinished_jobs:
+        unfinished_bios_jobs = \
+            [job.id for job in unfinished_jobs if _BIOS_JOB_NAME in job.name]
+        if unfinished_bios_jobs:
+            client.delete_jobs(job_ids=unfinished_bios_jobs)
+        else:
+            drac_job.validate_job_queue(node)
```
Comment (kpradip001) [inline ironic/drivers/modules/drac/management.py:203]: Just need discuss here one point, this logic will only clear pending bios jobs. In case of other pending jobs, validate_job_queue() will execute and raise an Unfinished config jobs found Error. 
So other pending jobs won't get deleted from job queue, Is this okay ?

**Comment by kpradip001** [comment]
> Looks pretty good. I just had one question on a test case. I can't quite figure out how it works...

I have addressed this in new commit, please take a look into this.
Thanks

File: ironic/drivers/modules/drac/management.py (hunk 12 above)
Comment (cdearborn) [reply ironic/drivers/modules/drac/management.py:203]: That's a good point Pradip!  I think the best thing to do is to preserve the prior behavior as much as possible, which was: if there are any unfinished config jobs then raise an exception.
I think we have 3 different cases here:
1. We have no unfinished config jobs
    In this case, we should just continue along so the cleaning step can run
2. We have only an unfinished bios config job
    In this case, I think we should delete it and continue so the cleaning step can run
3. We have more than 1 unfinished config job (this means that we have at least one non-bios config job)
    In this case, we should raise an exception and not delete any jobs
I think that taking this approach will minimally modify the prior behavior, which is probably a good thing.

File: ironic/tests/unit/drivers/modules/drac/test_management.py (hunk 11 above)
Comment (cdearborn) [reply ironic/tests/unit/drivers/modules/drac/test_management.py:460]: I'm still not sure how this works.  It looks like it was meant to test an exception that occurs in validate_job_queue() in the original code.  In the new code though, list_unfinished_jobs() will be called, and it will return no jobs since it is mocked above.  It will then fall through to set_boot_device() and never call validate_job_queue().  If validate_job_queue is never called, then setting it's side_effect below has no effect and should be removed.  What is raising the DracOperationError that is causing this test to pass?

File: ironic/tests/unit/drivers/modules/drac/test_management.py [hunk 13]
```diff
@@ -445,6 +466,100 @@ def test_set_boot_device_with_invalid_job_queue(
...
+                       'persistent': True}
+
+        mock__get_boot_device.return_value = boot_device
+        mock__get_next_persistent_boot_mode.return_value = 'IPL'
+
+        drac_mgmt.set_boot_device(self.node, ironic.common.boot_devices.DISK,
+                                  persistent=True)
+        mock_client.delete_jobs.assert_called_once_with(
+            job_ids=['JID_602553293345'])
+
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_management.py:542]: Let's add one more test case that tests list_unfinished jobs returning a bios job and another job

File: ironic/tests/unit/drivers/modules/drac/test_management.py [hunk 14]
```diff
@@ -445,6 +466,100 @@ def test_set_boot_device_with_invalid_job_queue(
...
+                       autospec=True)
+    @mock.patch.object(drac_mgmt, '_get_next_persistent_boot_mode',
+                       spec_set=True, autospec=True)
+    def test_set_boot_device_with_list_unfinished_jobs(
+            self, mock__get_next_persistent_boot_mode, mock__get_boot_device,
+            mock_validate_job_queue, mock_list_unfinished_jobs,
+            mock_get_drac_client):
+        mock_client = mock.Mock()
+        mock_get_drac_client.return_value = mock_client
+
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_management.py:519]: This is a nice test!

File: ironic/tests/unit/drivers/modules/drac/test_management.py [hunk 15]
```diff
@@ -445,6 +466,100 @@ def test_set_boot_device_with_invalid_job_queue(
...
+                       autospec=True)
+    @mock.patch.object(drac_mgmt, '_get_boot_device', spec_set=True,
+                       autospec=True)
+    @mock.patch.object(drac_mgmt, '_get_next_persistent_boot_mode',
+                       spec_set=True, autospec=True)
+    def test_set_boot_device_with_validate_job_queue(
+            self, mock__get_next_persistent_boot_mode, mock__get_boot_device,
+            mock_list_unfinished_jobs, mock_validate_job_queue,
+            mock_get_drac_client):
+        mock_client = mock.Mock()
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_management.py:461]: This is a nice test!

File: ironic/tests/unit/drivers/modules/drac/test_management.py (hunk 11 above)
Comment (cdearborn) [reply ironic/tests/unit/drivers/modules/drac/test_management.py:460]: So with the 2 test cases below, and if you add one more that tests returning multiple jobs from list_unfinished_jobs, I'm thinking that this test case could be deleted.

Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/drac/test_management.py:460]: Okay, so I will delete existing test case - test_set_boot_device_with_invalid_job_queue . And will add another test case for multiple jobs from list_unfinished_jobs.

File: ironic/tests/unit/drivers/modules/drac/test_management.py (hunk 13 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/drac/test_management.py:542]: Thanks, will add it.

File: ironic/drivers/modules/drac/management.py (hunk 12 above)
Comment (kpradip001) [reply ironic/drivers/modules/drac/management.py:203]: Okay, so as per current logic, case 1 & 2 are working.
For case 3 need to replace if statement with 
"if unfinished_bios_jobs and len(unfinished_jobs) == len(unfinished_bios_jobs) :" , am I right?

File: ironic/tests/unit/drivers/modules/drac/test_management.py (hunk 14 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/drac/test_management.py:519]: Thanks!

File: ironic/tests/unit/drivers/modules/drac/test_management.py (hunk 15 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/drac/test_management.py:461]: Thanks!

File: ironic/drivers/modules/drac/management.py (hunk 12 above)
Comment (cdearborn) [reply ironic/drivers/modules/drac/management.py:203]: Yes, that will correctly pick out the case where there are only bios bios config jobs, and in that case, we should delete them.  In practice, it should only be possible to have unfinished bios config job at a time, but i think it's a good idea to have the code work for more than 1.
The else clause will then execute for case #3.

Comment (kpradip001) [reply ironic/drivers/modules/drac/management.py:203]: Okay, will add this change, thanks!

Comment (kpradip001) [reply ironic/drivers/modules/drac/management.py:203]: Done

File: ironic/tests/unit/drivers/modules/drac/test_management.py (hunk 11 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/drac/test_management.py:460]: Done

File: ironic/tests/unit/drivers/modules/drac/test_management.py (hunk 13 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/drac/test_management.py:542]: I have added test_set_boot_device_with_multiple_unfinished_jobs test case, please review this.

**Review by cdearborn — APPROVED** [review]
Looks good!  Please go ahead and submit this patch upstream to Ironic.

**Comment by kpradip001** [comment]
I have created upstream patch in ironic, link: https://review.opendev.org/#/c/674021/

**Comment by cdearborn** [comment]
Closing this PR since it has been submitted upstream
